src/primitives/hrp.rs

```diff
@@ -119,6 +119,80 @@ impl Hrp {
         Ok(new)
     }
 
+    /// Parses the human-readable part from an object which can be formatted.
+    ///
+    /// The formatted form of the object is subject to all the same rules as [`Self::parse`].
+    /// This method is semantically equivalent to `Hrp::parse(&data.to_string())` but avoids
+    /// allocating an intermediate string.
+    pub fn parse_display<T: core::fmt::Display>(data: T) -> Result<Self, Error> {
+        use Error::*;
+
+        struct ByteFormatter {
+            arr: [u8; MAX_HRP_LEN],
+            index: usize,
+            error: Option<Error>,
+        }
+
+        impl core::fmt::Write for ByteFormatter {
+            fn write_str(&mut self, s: &str) -> fmt::Result {
+                let mut has_lower: bool = false;
+                let mut has_upper: bool = false;
+                for ch in s.chars() {
+                    let b = ch as u8; // cast ok, `b` unused until `ch` is checked to be ASCII
+
+                    // Break after finding an error so that we report the first invalid
+                    // character, not the last.
+                    if !ch.is_ascii() {
+                        self.error = Some(Error::NonAsciiChar(ch));
+                        break;
+                    } else if !(33..=126).contains(&b) {
+                        self.error = Some(InvalidAsciiByte(b));
+                        break;
+                    }
+
+                    if ch.is_ascii_lowercase() {
+                        if has_upper {
+                            self.error = Some(MixedCase);
+                            break;
+                        }
+                        has_lower = true;
+                    } else if ch.is_ascii_uppercase() {
+                        if has_lower {
+                            self.error = Some(MixedCase);
+                            break;
+                        }
+                        has_upper = true;
+                    };
+                }
+
+                // However, an invalid length error will take priority over an
+                // invalid character error.
+                if self.index + s.len() > self.arr.len() {
+                    self.error = Some(Error::TooLong(self.index + s.len()));
+                } else {
+                    // Only do the actual copy if we passed the index check.
+                    self.arr[self.index..self.index + s.len()].copy_from_slice(s.as_bytes());
+                }
+
+                // Unconditionally update self.index so that in the case of a too-long
+                // string, our error return will reflect the full length.
```

Comment:
nit: This comment is stale now I believe.

Reply:
We could just check the length at the start and early return.

Reply:
I don't see why this is stale. The line is called every time the `write_str` function is called, for the reason that the comment describes.

We could do a length check at the start and early return but I don't see what that would accomplish other than increasing the cyclotomic complexity of the function.

Reply:
ooo, I see. Don't mind me.